utils.go

```diff
@@ -16,18 +17,28 @@ const (
 	TagsHTTPHeaderPrefix = "Open-Tracing-Trace-Tags-"
 )
 
-// PropagateSpanInHeader encodes Span `ctx` to `h` as a series of
-// HTTP headers. Values are URL-escaped.
-func PropagateSpanInHeader(
+// InjectSpanInHeader encodes Span `sp` in `h` as a series of HTTP headers.
+// Values are URL-escaped.
+func InjectSpanInHeader(
 	sp Span,
 	h http.Header,
-	propagator SpanPropagator,
-) {
-	contextIDMap, tagsMap := propagator.PropagateSpanAsText(sp)
-	for headerSuffix, val := range contextIDMap {
+) error {
+	// First, look for a PROPAGATION_FORMAT_GO_HTTP_HEADER injector (our preference).
+	injector := sp.PropagationInjectorForFormat(PROPAGATION_FORMAT_GO_HTTP_HEADER)
```

Comment:
@michaelsembwever You didn't miss the comment, it wasn't there yet :)  [side note: Github would do well to fix their pull request UX -- it's damned confusing when you're trying to catch up after a day (or night) away from the firehose of individual updates]

So perhaps we are actually "in violent agreement", at least for the most part?? I hope so, anyway.

As for the "carriers": the motivating question is how the OpenTracing user who is dealing with process-boundary concerns (i.e., the person writing code like the lines we're commenting about here) is expected to pass in the `http.Header` or `map[string]string` (or whatever) that they're trying to inject into. I know that you wanted to do this – in Java, that is – by having the Injector rely on introspection. So, to be clear, something like this:

```
func InjectSpanInHeader(
	sp Span,
	h http.Header,
) error {
	// First, look for a GoHTTPHeader injector (our preference).

	//////////////////////////////////////////////////////////////////////////////
	// NOTE: we pass the http.Header *instance* in, so there's no need to pass
	// it again via a carrier.
	//////////////////////////////////////////////////////////////////////////////
	injector := sp.Injector(myHTTPHeader)
	if injector != nil {
		return injector.InjectSpan(sp)
	}

	// Else, fall back on SplitText.

	//////////////////////////////////////////////////////////////////////////////
	// NOTE: we pass both text maps in, so there's no need to pass them again
	// via a carrier.
	//////////////////////////////////////////////////////////////////////////////
	tracerState := map[string]string{}
	traceAttrs := map[string]string{}
	if injector = sp.Injector(tracerState, traceAttrs); injector == nil {
		return errors.New("No suitable injector")
	}
	if err := injector.Inject(sp); err != nil {
		return err
	}
	for headerSuffix, val := range tracerState {
		h.Add(ContextIDHTTPHeaderPrefix+headerSuffix, url.QueryEscape(val))
	}
	for headerSuffix, val := range traceAttrs {
		h.Add(TagsHTTPHeaderPrefix+headerSuffix, url.QueryEscape(val))
	}
	return nil
}
```

The above feels sketchy in Go, though it can be made to work. I preferred the security/certainty of the intermediary Injector/Extractor interfaces, personally. But this is not a huge thing for me. If we were to go with something like the above, I would just assume shorten this:

```
	if injector = sp.Injector(tracerState, traceAttrs); injector == nil {
		return errors.New("No suitable injector")
	}
	if err := injector.Inject(sp); err != nil {
		return err
	}
```

... to this:

```
	if err = sp.Inject(tracerState, traceAttrs); err != nil {
		return err
	}
```

That is, get rid of the `Injector` interface, and just make `Span.Inject()` more complex for the implementors. At least it could be more efficient that way. (Minor point: `Span.Inject()` would need variadic parameters if we want to get rid of the carrier types, which are basically just included for clarity)